### PR TITLE
#170947549 reset password feature

### PR DESCRIPTION
**Pull request by izzett222**
#### What does this PR do?
Adds the user password reset functionality
#### Description of Task to be completed?
- add the forgot password endpoint
- add the reset password endpoint
- configured SendGrid to send email
- add the verify token function
#### How should this be manually tested?
- clone this repo
- run npm install
- set up the database(follow the readme)
-  use the forgot password route and follow the instruction
- check your email 
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
#170947549
#### Screenshots (if appropriate)
Screen Shot 2020-02-18 at 20 31 38
Screen Shot 2020-02-18 at 20 32 56
Screen Shot 2020-02-18 at 20 42 44
Screen Shot 2020-02-18 at 20 45 08
Screen Shot 2020-02-18 at 20 44 37
#### Questions:

### CODE REVIEW

File: src/services/localesServices/locales/en.json [hunk 1]
```diff
@@ -22,6 +22,19 @@
...
+	"you are not allowed to perform this action": "you are not allowed to perform this action",
+	"token must be provided and valid": "token must be provided and valid",
 	"Lastname must be atleast 4 characters": "Lastname must be atleast 4 characters",
-	"Firstname must be atleast 4 characters": "Firstname must be atleast 4 characters"
-}
+	"email intro": "welcome back to barefoot nomad",
+	"email instruction": "To reset your password click here:",
+	"button text": "Reset your password",
+	"The allowable roles are manager, travel team member, requester, travel administrator, super administrator": "The allowable roles are manager, travel team member, requester, travel administrator, super administrator"
+}
```
Comment (Baraka-Mugisha) [inline src/services/localesServices/locales/en.json:45]: add newline please

Comment (izzett222) [reply src/services/localesServices/locales/en.json:45]: let me add it

File: src/tests/resetPassword.test.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    chai
+      .request(index)
+      .put(`/api/v1/auth/resetpassword?token=${notSignupToken}`)
+      .send({
+        password: '@fnfrjfrA',
+      })
+      .end((err, res) => {
+        expect(res.status).to.equal(403);
+        expect(res.body.status).to.equal(403);
+        expect(res.body.message).to.equal('you are not allowed to perform this action');
```
Comment (Baraka-Mugisha) [inline src/tests/resetPassword.test.js]: could you change res.body.message to res.body.error as other features?

File: src/middlewares/verifyUser.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    req.user = user;
+    return next();
+  } catch (err) {
+    if (err.name === 'JsonWebTokenError') {
+      return res.status(400).json({
+        status: 400,
+        message: res.__('token must be provided and valid'),
+      });
+    }
+    return res.status(500).json({
```
Comment (Baraka-Mugisha) [inline src/middlewares/verifyUser.js]: could you use responseHandler.js to return the response?

File: src/services/localesServices/locales/en.json (hunk 1 above)
Comment (izzett222) [reply src/services/localesServices/locales/en.json:45]: done

File: src/tests/resetPassword.test.js (hunk 2 above)
Comment (izzett222) [reply src/tests/resetPassword.test.js]: I changed the response to error

File: src/middlewares/verifyUser.js (hunk 3 above)
Comment (izzett222) [reply src/middlewares/verifyUser.js]: done

**Review by Baraka-Mugisha — DISMISSED** [review]

**Comment by bbaime98** [comment]
could you add a screenshot of on success and failure?

File: src/utils/verifyToken.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import jwt from 'jsonwebtoken';
+import dotenv from 'dotenv';
+
+dotenv.config();
```
Comment (Mnickii) [inline src/utils/verifyToken.js]: can't we use the decode function in the token handler file? you can rename it to verify token

File: src/utils/send-reset-email.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+const hostUrl = `http://${process.env.BASE_URL}`;
+const mailGenerator = new mailgen({
+  theme: 'default',
+  product: {
+    name: 'barefoot nomad',
+    link: 'barefoot nomad'
+  }
+});
+const sendMsg = (newUserEmail, token, firstName, content) => {
+  const email = mailGenerator.generate({
```
Comment (Mnickii) [inline src/utils/send-reset-email.js]: Let's have a single way to have our emails look like @Ntare22 @izzett222

**Review by Mnickii — CHANGES_REQUESTED** [review]

File: src/utils/verifyToken.js (hunk 4 above)
Comment (izzett222) [reply src/utils/verifyToken.js]: the decode function is a middleware function

File: src/utils/send-reset-email.js (hunk 5 above)
Comment (izzett222) [reply src/utils/send-reset-email.js]: we changed the send email function and the email are similar

**Review by Baraka-Mugisha — DISMISSED** [review]

**Review by Ntare22 — DISMISSED** [review]
LGTM

File: src/utils/verifyToken.js (hunk 4 above)
Comment (Mnickii) [reply src/utils/verifyToken.js]: then let's move this function into the token handler and call it in the decode middleware

File: src/utils/send-reset-email.js (hunk 5 above)
Comment (Mnickii) [reply src/utils/send-reset-email.js]: Let's also refactor the welcome email to this format

Comment (izzett222) [reply src/utils/send-reset-email.js]: we already changed the welcome email that Jim created by using the new send email function

File: src/utils/verifyToken.js (hunk 4 above)
Comment (izzett222) [reply src/utils/verifyToken.js]: let me add it to the middleware

Comment (izzett222) [reply src/utils/verifyToken.js]: done, added verifyToken to the decode middleware

**Review by Baraka-Mugisha — DISMISSED** [review]

**Review by Mnickii — DISMISSED** [review]

**Review by Mnickii — APPROVED** [review]

**Review by bbaime98 — APPROVED** [review]

**Review by Baraka-Mugisha — APPROVED** [review]

**Pull request merged**